rs/tests/driver/src/driver/bootstrap.rs

```diff
@@ -567,13 +518,111 @@ fn node_to_config(node: &Node) -> NodeConfiguration {
     }
 }
 
-fn configure_setupos_image(
+#[derive(Clone)]
+pub enum NestedVersionTarget {
+    Mainnet,
+    /// When true, use the branch -test variant.
+    Branch(bool),
+    Published {
+        version: String,
+        url: Url,
+        sha256: String,
+    },
+}
+
+// Setup nested VMs. NOTE: This is different from `setup_and_start_vms` in that
+// we need to configure and push a SetupOS image for each node.
+pub fn setup_nested_vms(
+    nodes: &[NestedNode],
+    env: &TestEnv,
+    farm: &Farm,
+    group_name: &str,
+    nns_url: &Url,
+    nns_public_key: &str,
+    to: &NestedVersionTarget,
+) -> anyhow::Result<()> {
+    let mut join_handles: Vec<JoinHandle<anyhow::Result<()>>> = vec![];
+    for node in nodes {
+        let t_farm = farm.to_owned();
+        let t_env = env.to_owned();
+        let t_group_name = group_name.to_owned();
+        let t_vm_name = node.name.to_owned();
+        let t_nns_url = nns_url.to_owned();
+        let t_nns_public_key = nns_public_key.to_owned();
+        let t_to = to.to_owned();
```

Comment:
nit: By using [thread::scope](https://doc.rust-lang.org/stable/std/thread/fn.scope.html), you could avoid creating all these owned copies and just use the references directly (if I checked correctly, the called functions only require shared refs). In addition, it ensures that every thread in the scope is joined in the Scope's destructor.

Reply:
🙌